nmdc_runtime/site/changesheets/scripts/missing_neon_soils_ecosystem_data.py

```diff
@@ -0,0 +1,338 @@
+@click.option(
+    "--apply_changes", is_flag=True, default=False, help=("Apply the changes")
+)
+@click.option("--use_dev_api", is_flag=True, default=False, help=("Use the dev "
+                                                              "API"))
+def generate_changesheet(study_id, apply_changes, use_dev_api):
+    """
+    Generate a changesheet for missing ecosystem data for NEON soils samples by:
+    1. Retrieving GOLD biosamples for the given study
+    2. Finding the corresponding NMDC biosamples and omics processing records for each GOLD biosample
+    3. Comparing the GOLD biosample to the NMDC biosamples and omics processing records
+    4. Generating a changesheet for the differences
+
+    WARNING: This script is not idempotent. It will generate a new changesheet each time it is run.
+
+    Changesheet is written to nmdc_runtime/site/changesheets/changesheets_output
+
+    :param study_id: The GOLD study ID
+    :param apply_changes: Whether to apply the changes via the
+    metadata/changesheets:apply endpoint (default: False) NOTE: not implemented!
+    :param use_dev_api: Use the dev API (default: False)
+    :return:
+    """
+    start_time = time.time()
+    logging.info("starting missing_neon_soils_ecosystem_data.py...")
+    logging.info(f"study_id: {study_id}")
+
+    gold_api_client = GoldApiClient(
+        base_url=os.getenv("GOLD_API_BASE_URL"),
+        username=os.getenv("GOLD_API_USERNAME"),
+        password=os.getenv("GOLD_API_PASSWORD"),
+    )
+    logging.info("connected to GOLD API...")
+
+    if use_dev_api:
+        base_url = os.getenv("API_HOST_DEV")
+        logging.info("using dev API...")
+    else:
+        base_url = os.getenv("API_HOST")
+        logging.info("using prod API...")
+
+    runtime_api_user_client = RuntimeApiUserClient(
+        base_url=base_url,
+        username=os.getenv("API_QUERY_USER"),
+        password=os.getenv("API_QUERY_PASS"),
+    )
+    logging.info("connected to NMDC API...")
+
+    # Retrieve GOLD biosamples for the given study
+    gold_biosamples = gold_api_client.fetch_biosamples_by_study(study_id)
+    logging.info(f"retrieved {len(gold_biosamples)} biosamples from GOLD API...")
+
+    # omics processing to biosamples map generated by a SPARQL query
+    omprc_to_bs_map = read_omics_processing_to_biosample_map()
+
+    changesheet = Changesheet(name=NAME)
+    # For each GOLD biosample, find the corresponding NMDC biosamples
+    nmdcbs_count = 0
+    unfindable_count = 0
+    unfindable_goldbs_id = []
+    for goldbs in gold_biosamples:
+        (
+            nmdc_biosamples,
+            omprc_records,
+        ) = gold_biosample_to_nmdc_biosamples_and_omics_processing_records(
+            runtime_api_user_client, omprc_to_bs_map, goldbs
+        )
+        if not nmdc_biosamples:
+            logging.warning(
+                f"no corresponding NMDC biosamples found for {goldbs['biosampleGoldId']}..."
+            )
+            unfindable_count += 1
```

Comment:
I think this is always the length of `unfindable_goldbs_id` (one line later) and could be computed on-demand (in the `logging.info` statement below).

Reply:
done